### DISCUSSION TITLE
Match strings that start and end with the same character

### DISCUSSION THREAD

**Discussion by matias-santurio** [pest-parser/pest#894] [Q&A]
I'm trying to write a grammar rule that matches strings that start and end with the same character, so a rule that matches:

```
aba
22
zhelloz
```

but not

```
abb
21
```

My first thought was to write something like

```
TEST = {PUSH(ANY) ~ ANY* ~ POP }
```

but of course, matching is greedy so `ANY*` consumes the whole string, is there a way to get around this?

**Comment by Tartasprint** [comment]
You can avoid consuming the whole string using predicates:
```pest
all_but_last = {(!(ANY ~ EOI) ~ ANY)*}
```
But I'm not sure that this method is very efficient (I think that for every character the parser checks if the next is EOI). Using that I was able to write the grammar you described, see [this grammar on the playground](https://pest.rs/?g=N4Ig5gTghgtjURALhBAzgAgLweACggB8BLAOwDcoAbYgEwH0bSBTASgCoBfAHVLMpoMmzLAAF8AQjzNSDAPYAzRmTYA-AIIA5AJodVM%2BUuE8K1OspbYM4gAoBVAMoAJPFt0ZVGPFLf7Z9RQs1Nw4PDBsAeRswgwCjFRNYwOErYAxuEG5eDIwTEAAaEDIABwBXABdkECgAC2YqKjkoXlr6xtJeAEYAJk7s0hBOIA#editor).
